R/tlf_ae_specific.R

```diff
@@ -67,57 +68,182 @@ tlf_ae_specific <- function(outdata,
+  )
+
+  rwidth_2_within <- rep(1, length(col_tbl_within) * n_group)
+
+  rwidth_1_within <- tapply(
+    rwidth_2_within,
+    c(rep(1:n_group, each = length(col_tbl_within))),
+    sum
   )
 
-  # TODO: add logic for CI and p-value with multipel groups following WMA mock up table.
-  # colheader_ci <- c(paste("Difference in % vs", group[reference_group]),
-  # "Estimate (95% CI)")
+  colborder_within <- rep(colborder_within, n_group)
 
-  # colheader_p <- c("", "p-value")
-  # colheader <- paste(colheader_n, colheader_ci, colheader_p, sep = " | ")
+  # between (comparison statistics)
 
-  colheader <- colheader_n
+  col_tbl_between <- strsplit(names(tbl), "_") |>
+    unlist() |>
+    (\(list) list[list %in% c("diff", "ci", "p")])() |>
+    unique()
+
+  if (length(col_tbl_between) > 0) {
+    colhead_between <- paste(
+      sapply(
+        X = col_tbl_between,
+        FUN = switch,
+        "diff" = "Estimate",
+        "ci" = paste0("(", outdata$ci_level * 100, "% CI)"),
+        "p" = "p-value",
+      ),
+      collapse = " | "
+    )
+
+    colhead_1_between <- paste("Difference in %",
+      outdata$group[between_total[-reference_group]],
+      "vs.",
+      outdata$group[reference],
+      collapse = " | "
+    )
+
+    colhead_2_between <- paste(rep(colhead_between, n_comparisons), collapse = " | ")
+
+    colborder_between <- sapply(
+      X = col_tbl_between,
+      FUN = switch,
+      "diff" = "single",
+      "ci" = "",
+      "p" = "single",
+      USE.NAMES = FALSE
+    )
+
+    rwidth_2_between <- rep(1, length(col_tbl_between) * n_comparisons)
+
+    rwidth_1_between <- tapply(
+      rwidth_2_between,
+      c(rep(1:n_comparisons, each = length(col_tbl_between))),
+      sum
+    )
+
+    colborder_between <- rep(colborder_between, n_comparisons)
+  } else {
+    colhead_between <- colhead_1_between <- colhead_2_between <- NULL
+    rwidth_1_between <- rwidth_2_between <- colborder_between <- NULL
+  }
+
+  # Column headers
+
+  colheader <- c(
+    paste0(" | ", paste0(c(colhead_1_within, colhead_1_between), collapse = " | ")),
+    paste0(" | ", paste0(c(colhead_2_within, colhead_2_between), collapse = " | "))
+  )
 
   # Relative width
 
   if (is.null(col_rel_width)) {
-    rel_width <- c(3, rep(1, 2 * n_group))
+    rwidth_2 <- c(3, rwidth_2_within, rwidth_2_between)
+    rwidth_1 <- c(3, rwidth_1_within, rwidth_1_between)
   } else {
-    rel_width <- col_rel_width
+    rwidth_2 <- col_rel_width
+
+    rw_1_recalc_w <- tapply(
+      col_rel_width[2:(n_group * length(col_tbl_within) + 1)],
+      c(rep(1:n_group, each = length(col_tbl_within))), sum
+    )
+
+    rw_1_recalc_b <- if (length(col_tbl_between) > 0) {
+      tapply(
+        col_rel_width[(n_group * length(col_tbl_within) + 2):n_col],
+        c(rep(1:n_comparisons, each = length(col_tbl_between))), sum
+      )
+    } else {
+      NULL
+    }
+
+    rwidth_1 <- c(
+      rwidth_2[1],
+      rw_1_recalc_w,
+      rw_1_recalc_b
+    )
```

Comment:
not thrilled with this control flow and needing to re-calc the top-level widths. please let me know if you have a more elegant way to do this.